main.cpp

```diff
@@ -0,0 +1,60 @@
+#include <iostream>
+#include <thread>
+#include <vector>
+#include <future>
+
+std::mutex mtx;
+std::condition_variable cv;
+bool finished = false;
+
+class Producer {
+private:
+    std::thread::id thread_id;
+    double value;
+
+public:
+    Producer(std::thread::id id, double val) : thread_id(id), value(val) {}
+
+    void operator()() {
+        std::unique_lock<std::mutex> lock(mtx);
+        while (value != 0) {
+            std::cout << std::hash<std::thread::id>{}(thread_id) << " sent: " << value << std::endl;
+            value /= 10;
+        }
+        std::cout << std::hash<std::thread::id>{}(thread_id) << " finished" << std::endl;
+
+        if (finished) {
+            cv.notify_one();
+        }
+        else {
+            finished = true;
+            cv.wait(lock);
+            cv.notify_one();
+        }
+    }
+};
+
+
+int main() {
+    std::vector<std::future<void>> producers;
+    std::thread consumer([]() {});
```

Comment:
The consumer was supposed to get the messages and print them. It definitely wasn't supposed to do nothing..